### PR TITLE
prov/tcp: Add rendezvous protocol support

### PR DESCRIPTION
**Pull request by shefty**
A set of changes identified while preparing rendezvous support.  Remove unused dynamic receive buffer code from tcp and rxm (no longer supported by the tcp provider).  Add an optimized indexer abstraction that targets the tcp rendezvous implementation.  Etc.

Rendezvous support is not included in these changes.  That is still a work in progress.  Just trying to break up the change set.  Unfortunately, until rendezvous is complete, some of these changes will not be tested.

### CODE REVIEW

**Comment by zachdworkin** [comment]
CI will fail until you have this PR added into your tree:
https://github.com/ofiwg/libfabric/pull/9186
specifically 3385d96619e58bc3c5852ac50c915067e094c7fe

**Comment by shefty** [comment]
@ooststep - I've added the partial rendezvous protocol patch for review

**Review by ooststep — COMMENTED** [review]
code looks good. I'll start looking at validating.

do we need to worry about protocol backward compatibility?

File: prov/tcp/src/xnet_progress.c [hunk 1]
```diff
@@ -730,27 +824,76 @@ static int xnet_op_tagged(struct xnet_ep *ep)
...
-		}
-		return -FI_EAGAIN;
+		FI_WARN(&xnet_prov, FI_LOG_EP_DATA, "Invalid cts index\n");
+		return -FI_EINVAL;
 	}
 
-start:
-	return xnet_start_recv(ep, rx_entry);
+	memcpy(&rx_entry->hdr, &msg->hdr, (size_t) msg->hdr.base_hdr.hdr_size);
+//	rx_entry->hdr.base_hdr.hdr_size.size = msg->hdr.base_hdr.hdr_size;
```
[the remark this answers is not in the document]
Comment (shefty) [reply prov/tcp/src/xnet_progress.c:884]: The last patch is still a work in progress that's not ready or cleaned up.  We do have to worry about protocol compatibility.  This flow won't be hit unless the max saved size is reduced, but I have a TODO to see how to enable the flow programmatically.  I don't have peek, claim, discard implemented yet, but most everything else should be there.  Nothing has been tested.

[the remark this answers is not in the document]
Comment (shefty) [reply prov/tcp/src/xnet_progress.c]: All patches should be cleaned up now.  The changes still aren't tested (other than making sure things still work when rendezvous isn't enabled).  I'm pulling them to test now.

[the remark this answers is not in the document]
Comment (shefty) [reply prov/tcp/src/xnet_progress.c]: PR updated with 2 minor fixes.  I verified basic functionality with fabtests forcing the rendezvous protocol.  To enable rendezvous (for now), reduce the FI_TCP_MAX_SAVED_SIZE (double check actual name) to something like 16k.  16k is the eager buffer size, but the value doesn't have to be aligned to that.  However, rendezvous won't be used for transfers smaller than the eager size.

Still to do: test peek/claim/discard cases, forcing specific corner cases, handle wire protocol compatibility, and analyze performance.

File: prov/tcp/src/xnet_progress.c [hunk 2]
```diff
@@ -183,6 +179,69 @@ static int xnet_handle_truncate(struct xnet_ep *ep)
...
+		 struct xnet_xfer_entry *rx_entry)
+{
+	struct xnet_conn *conn;
+	uint8_t cts_ctx;
+	int ret;
+
+	assert(xnet_progress_locked(xnet_rdm2_progress(rdm)));
+	if (!ep) {
+		if (xnet_get_conn(rdm, rx_entry->src_addr, &conn)) {
+			FI_WARN(&xnet_prov, FI_LOG_EP_DATA,
```
Comment (ooststep) [inline prov/tcp/src/xnet_progress.c:220]: compiler complained ret may be uninitialized here

Comment (shefty) [reply prov/tcp/src/xnet_progress.c:220]: uhm, yeah, that's true.  I'll update.  I don't expect to ever go through this path, but it's theoretically possible

File: prov/tcp/src/xnet_progress.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (shefty) [reply prov/tcp/src/xnet_progress.c]: update: peek doesn't work - debugging it

File: prov/tcp/src/xnet_srx.c [hunk 3]
```diff
@@ -203,7 +203,7 @@ static int xnet_check_match(const union xnet_hdrs *hdr,
 {
 	uint64_t cur_tag;
 
-	if (hdr->base_hdr.op != ofi_op_tagged)
+	if (hdr->base_hdr.op != xnet_op_tag && hdr->base_hdr.op == xnet_op_tag_rts)
```
Comment (shefty) [inline prov/tcp/src/xnet_srx.c:206]: The main problem comes from using ```==``` instead of ```!=``` above...

**Comment by shefty** [comment]
@ooststep - This is now passing fabtests with rendezvous flow enabled.

**Comment by shefty** [comment]
Intel CI failure is new sighting in the udp provider and unrelated to this PR.

**Comment by shefty** [comment]
Added patch that negotiates rendezvous support with peer.  Because this requires an update to the cm messages in a backwards compatible manner, rendezvous is only enabled for rdm endpoints (which uses its own cm messages).  The cm messages for msg endpoints is more difficult to deal with, as mismatched cm versions are rejected, not negotiated down.

Last patch forces the use of rendezvous for all messages larger than 256 bytes.  This is for CI testing purposes only.  That patch will not be merged (intentionally fails the DCO check to prevent merging).

**Comment by shefty** [comment]
Intel CI failure: ```fi_multinode -n 128 -s cb17-eth2 -p 'tcp' -C msg -z full_mesh -I 1 -T```.

Looks like test just hung?  I'm investigating.  Other tcp tests, but it looks like MPI testing may be using tcp;rxm, instead of tcp alone.  Investigating that as well.

**Comment by shefty** [comment]
@ooststep - The latest changes are passing CI, except for multinode.  (Note that I force rendezvous at a ridiculously small size.)  I'm investigating the multinode failure.

**Comment by shefty** [comment]
bot:aws:retest
